File: 383-ransom-note/ransom-note.java
class Solution {
    public boolean canConstruct(String ransomNote, String magazine) {
        // // vector wala hai 
        // int[] magazineFreq = new int[128]; // Fixed size java array
        
        // // Loop 1: Magazine inventory update
        // for (int i = 0; i < magazine.length(); i++) {
        //     char magChar = magazine.charAt(i);
        //     magazineFreq[magChar]++;
        // }
        
        // // Loop 2: Ransom Note verification from magazineFreq array
        // for (int i = 0; i < ransomNote.length(); i++) {
        //     char noteChar = ransomNote.charAt(i);
            
        //     if (magazineFreq[noteChar] == 0) {
        //         return false;
        //     }
        //     magazineFreq[noteChar]--;
        // }
        
        // return true;

        // hashmap se hua hai

        HashMap<Character, Integer> magazineFreq = new HashMap<>(); // Array hata kar map banaya
        
        // Loop 1: On-the-fly key-value insertion
        for (int i = 0; i < magazine.length(); i++) {
            char magChar = magazine.charAt(i);
            magazineFreq.put(magChar, magazineFreq.getOrDefault(magChar, 0) + 1);
        }
        
        // Loop 2: Checking using map methods
        for (int i = 0; i < ransomNote.length(); i++) {
            char noteChar = ransomNote.charAt(i);
            
            // Map se value nikali (agar key nahi mili toh default 0 milega)
            if (magazineFreq.getOrDefault(noteChar, 0) == 0) {
                return false;
            }
            // Stock mein se ek piece kam kiya
            magazineFreq.put(noteChar, magazineFreq.get(noteChar) - 1);
        }
        
        return true;
    }
}

File: 383-ransom-note/ransom-note.cpp
class Solution {
public:
    bool canConstruct(string ransomNote, string magazine) {
        // Vector hata kar asli unordered_map lagaya snake_case mein
        unordered_map<char, int> magazine_freq; 
        
        // Loop 1: Dynamic map insertion
        for (int i = 0; i < magazine.length(); i++) {
            char mag_char = magazine[i]; 
            magazine_freq[mag_char]++; 
        }
        
        // Loop 2: Checking from map stock
        for (int i = 0; i < ransomNote.length(); i++) {
            char note_char = ransomNote[i]; 
            
            if (magazine_freq[note_char] == 0) {
                return false; 
            }
            magazine_freq[note_char]--; 
        }
        
        return true;
    }
};